Mathlib/CategoryTheory/Grothendieck.lean

```diff
@@ -306,6 +310,58 @@ def pre (G : D ⥤ C) : Grothendieck (G ⋙ F) ⥤ Grothendieck F where
   map_id X := Grothendieck.ext _ _ (G.map_id _) (by simp)
   map_comp f g := Grothendieck.ext _ _ (G.map_comp _ _) (by simp)
 
+section FunctorFrom
+
+variable {E : Type*} [Category E]
+
+variable (F) in
+/-- The inclusion of a fiber `F.obj c` of a functor `F : C ⥤ Cat` into its Grothendieck
+construction.-/
+@[simps]
+def ι (c : C) : F.obj c ⥤ Grothendieck F where
+  obj d := ⟨c, d⟩
+  map f := ⟨𝟙 _, eqToHom (by simp) ≫ f⟩
+  map_id d := by
+    dsimp
+    congr
+    simp only [Category.comp_id]
+  map_comp f g := by
+    apply Grothendieck.ext _ _ (by simp)
+    simp only [comp_base, ← Category.assoc, eqToHom_trans, comp_fiber, Functor.map_comp,
+      eqToHom_map]
+    congr 1
+    simp only [eqToHom_comp_iff, Category.assoc, eqToHom_trans_assoc]
+    apply Functor.congr_hom (F.map_id _).symm
+
+instance faithful_ι (c : C) : (ι F c).Faithful where
+  map_injective f := by
+    injection f with _ f
+    rwa [cancel_epi] at f
+
+variable (fib : ∀ c, F.obj c ⥤ E) (hom : ∀ {c c' : C} (f : c ⟶ c'), fib c ⟶ F.map f ⋙ fib c')
+variable (hom_id : ∀ c, hom (𝟙 c) = eqToHom (by simp only [Functor.map_id]; rfl))
+variable (hom_comp : ∀ c₁ c₂ c₃ (f : c₁ ⟶ c₂) (g : c₂ ⟶ c₃), hom (f ≫ g) =
+  hom f ≫ whiskerLeft (F.map f) (hom g) ≫ eqToHom (by simp only [Functor.map_comp]; rfl))
+
+/-- Construct a functor from `Grothendieck F` to another category `E` by providing a family of
+functors on the fibers of `Grothendieck F`, a family of natural transformations on morphisms in the
+base of `Grothendieck F` and coherence data for this family of natural transformations. -/
+@[simps]
+def functorFrom : Grothendieck F ⥤ E where
+  obj X := (fib X.base).obj X.fiber
+  map {X Y} f := (hom f.base).app X.fiber ≫ (fib Y.base).map f.fiber
+  map_id X := by simp [hom_id]
+  map_comp f g := by simp [hom_comp]
+
+/-- `Grothendieck.apply` and `Grothendieck.functorFrom` fulfill a kind of "β reduction", in which
+application and abstraction cancel each other out. -/
+theorem ι_functorFrom (c : C) : ι F c ⋙ (functorFrom fib hom hom_id hom_comp) = fib c := by
```

Comment:
We generally try to avoid non-definitional equality of functors. Is this equality unavoidably necessary in your application or would a natural isomorphism suffice?

Reply:
That's why I didn't make it a simp lemma. My gut feeling is that we'll need the equality but maybe it's also enough if it just reduces on objects, which it does... Maybe I'll just add the iso for now and add an equality option if we really need it later.